### DISCUSSION TITLE
Programmatically building Schema objects -- use "ObjectType" or "ObjectTypeDefinition" etc?

### DISCUSSION THREAD

**Discussion by GavinRay97** [graphql-java/graphql-java#2681] [Q&A] [answered]
I've been programmatically building schemas using `GraphQL.newObjectType()`, etc:

```kt
val schema = GraphQLSchema.newSchema().query(
  GraphQLObjectType.newObject().name("Query").fields() // ...
)
```

But I see there are a distinction between IE, `GraphQLObjectType` and `GraphQLObjectTypeDefinition` (ditto for all other types), and that the `-TypeDefinition` nodes are the ones used to build `TypeRegistry`.

What's the difference between these two, and which should I be using to dynamically generate schemas programmatically?

Thank you =)

**Comment by bbakerman** [comment]
We have 2 phases in the definition of schema.  The raw Abtract Syntax Tree objects are created by the parser during textual analysis.

```
type Foo implements Bar {
   f : String
}
```

The above SDL text would result in a parsed `graphql.language.ObjectTypeDefinition`, which is the AST elements for that object type.  Its valid syntax but its and invalid schema (since there is no `Query` type not does the interface `Bar` exist.)

The second phase involves `graphql.schema.idl.SchemaGenerator` taking that `TypeRegistry` of parsed AST elements and make a valid `GraphqlSchema` full of runtime elements like `graphql.schema.GraphQLObjectType` and so on.  These are validated and turned into singleton objects with type references (circular references) resolved.

**Comment by bbakerman** [comment, marked as answer]
If you make a schema programmatically - then you don't need AST elements - you just need `raphql.schema.GraphQXXXX` classes

**Comment by GavinRay97** [comment]
Ahh understood, thank you!

So if I am interpreting this right, what I want to do is something like this?
(Assuming that the `GraphQLSchema` is being generated in another function that doesn't add code-registry datafetchers)

```java
GraphQLSchema.Builder schemaBuilder = GraphQLSchema.newSchema();

schemaBuilder.query(GraphQLObjectType.newObject()
        .name("Query")
        .field(GraphQLFieldDefinition.newFieldDefinition()
                .name("sayHello")
                .type(Scalars.GraphQLString)
                .argument(GraphQLArgument.newArgument()
                        .name("name")
                        .type(Scalars.GraphQLString)
                        .build())
        )
        .build());

GraphQLSchema schema = schemaBuilder.build();
schema.transform(builder -> {
    GraphQLCodeRegistry.Builder codeRegistryBuilder = GraphQLCodeRegistry.newCodeRegistry();

    codeRegistryBuilder
            .dataFetcher(
                    FieldCoordinates.coordinates("Query", "sayHello"),
                    (DataFetcher<String>) environment -> "Hello " + environment.getArgument("name")
            )
            .build();
});

```

**Reply by bbakerman** [reply]
Yes - but you can build the `GraphQLCodeRegistry` at the same time as the schema - no need for the `schema.transform`

Have  a look at https://www.graphql-java.com/documentation/schema

**Reply by GavinRay97** [reply]
Got it -- thank you =)
